### PR TITLE
Check owners of a Kubernetes type before validation

### LINKED ISSUE #250: check for "Controlled By" section before admitting pods
Right now if we validate and admit a deployment, we validate the pods it creates again. 

It looks like if pods are created by other objects (Deployment, ReplicaSet etc) then there's a "Controlled By" section in the pod description which we could check for before validation. If the "Controlled By" section exists, we can let the pod through (assuming whatever created it was already validating and approved)

```shell
$ kubectl describe pod test-5d54fc85c-bn72b 
Name:           test-5d54fc85c-bn72b
Namespace:      default
Node:           gke-kritis-default-pool-f1c1e77c-n3sw/10.128.0.2
Start Time:     Wed, 15 Aug 2018 10:19:01 -0700
Labels:         pod-template-hash=181097417
                run=test
Annotations:    kubernetes.io/limit-ranger=LimitRanger plugin set: cpu request for container test
Status:         Running
IP:             10.64.2.52
Controlled By:  ReplicaSet/test-5d54fc85c
```

For this to work, we'll need to add:
1. ReplicaSet type to kritis, since Deployments create ReplicaSets which in turn create Pods
2. The actual logic for checking the "Controlled By" section

### PR DESCRIPTION
**Pull request by priyawadhwa**
This PR adds support for checking if an object's owner has been
validated before starting validation. If an object is owned by a type
supported by Kritis (currently Pods, Deployments and ReplicaSets) then
we know that it has already been validated and we don't have to validate
again.

This should fix #250

### CODE REVIEW

**Comment by dlorenc** [comment]
How safe is this? Can someone set an owner reference after the fact?

**Comment by priyawadhwa** [comment]
So it looks like you can set it after the fact, from [here](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents)
>You can also specify relationships between owners and dependents by manually setting the ownerReference field.

Theoretically you can set it to anything, even something that doesn't exist. If we make sure that:
1. An object's owner actually exists
2. The object contains the same images as the owner (or a subset of the owner's images)

would that be enough to allow an object?

**Comment by tejal29** [comment]
Is there a way to get the name of the owner object? If that is the case, your suggestion 1 to check  if owner exists should work. 
For pt 2, we need to think about images with tags.

File: pkg/kritis/admission/admission.go [hunk 1]
```diff
@@ -272,6 +285,22 @@ func unmarshalDeployment(r *http.Request) (*appsv1.Deployment, v1beta1.Admission
...
 }
 
+func checkOwners(meta *metav1.ObjectMeta) bool {
+	owners := meta.GetOwnerReferences()
+	if owners == nil {
+		return false
+	}
+	for _, o := range owners {
+		for _, s := range constants.SupportedTypes {
+			if o.Kind == s {
```
Comment (tejal29) [inline pkg/kritis/admission/admission.go]: add a check if onwer is running like we mentioned.

File: pkg/kritis/admission/admission.go [hunk 2]
```diff
@@ -178,11 +178,17 @@ func ReviewHandler(w http.ResponseWriter, r *http.Request) {
 }
 
 func reviewDeployment(deployment *appsv1.Deployment, ar *v1beta1.AdmissionReview) {
+	images := DeploymentImages(*deployment)
+	// check if the Deployments's owner has already been validated
+	if checkOwners(images, &deployment.ObjectMeta) {
+		return
```
Comment (tejal29) [inline pkg/kritis/admission/admission.go]: Can we log here and elsewhere so its easier to know why the object was admitted.

**Pull request merged**